File: implementations/javascript/binary_search.js
const SIZES = { small: 100, medium: 10000, large: 1000000 };
const n = SIZES[process.argv[2]];

function binarySearch(arr, n, target) {
    let lo = 0, hi = n - 1;
    while (lo <= hi) {
        const mid = lo + ((hi - lo) >> 1);
        if (arr[mid] === target) return mid;
        else if (arr[mid] < target) lo = mid + 1;
        else hi = mid - 1;
    }
    return -1;
}

const arr = new Array(n);
for (let i = 0; i < n; i++) arr[i] = i * 2;
const target = (n - 1) * 2;
console.log(binarySearch(arr, n, target));

File: implementations/python/binary_search.py
import sys

SMALL  = 100
MEDIUM = 10000
LARGE  = 1000000

SIZES = {"small": SMALL, "medium": MEDIUM, "large": LARGE}

def binary_search(arr, n, target):
    lo, hi = 0, n - 1
    while lo <= hi:
        mid = lo + (hi - lo) // 2
        if arr[mid] == target:
            return mid
        elif arr[mid] < target:
            lo = mid + 1
        else:
            hi = mid - 1
    return -1

if __name__ == "__main__":
    n = SIZES[sys.argv[1]]
    arr = [i * 2 for i in range(n)]
    target = (n - 1) * 2
    print(binary_search(arr, n, target))
